### PR TITLE
Add on_deactivate callback for luaentities

### PR DESCRIPTION
**Pull request by hecktest**
Inspired by @sorcerykid 's work in #9717

_Scripting API, maintenance_
**Ready for review**
This PR adds an `on_deactivate` event for entities, called when an entity is about to get removed or unloaded.
The function is called right **before** an entity is marked for removal - meaning methods like `ObjectRef:get_pos` are still safe to use within it, and won't return nil.
The event is the last chance to safely run user code before the object is considered "gone". Keep in mind that step functions must still check whether the object has been removed before them. (fixing that is a task for another day)

On the binary side, the act of marking entities for death by setting a bool on them was replaced by a function that does this properly and raises the aforementioned event.

**How to test**
Create a world in devtest, spawn a callback entity, whack it with a sword until it dies.

### CODE REVIEW

File: src/server/serveractiveobject.cpp [hunk 1]
```diff
@@ -73,3 +73,21 @@ void ServerActiveObject::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &
 		m_messages_out.pop();
 	}
 }
+
+void ServerActiveObject::markForRemoval()
+{
+	if (!m_pending_removal) {
+		if (!isGone())
```
Comment (hecktest) [inline src/server/serveractiveobject.cpp:80]: This is to ensure that on_deactivated is not called twice on the same entity, without going too deep into the intricacies of when and how the remove/deactivate flags interact with one another.

**Comment by runsy** [comment]
Cool! I need it for my petz mod. I need to save some petz data when entity deactivated, currently it is done saving over and over the data each second. This PR will be a great advance.

**Comment by hecktest** [comment]
This might not get called if the server crashes, so you should periodically be saving data anyway.

File: src/server/serveractiveobject.h [hunk 2]
```diff
@@ -220,6 +226,7 @@ class ServerActiveObject : public ActiveObject
 		  it could be confused to some other object by some client.
 		- This is usually set to true by the step() method when the object wants
 		  to be deleted but can be set by anything else too.
+		Note: Do not assign this directly, use markForRemoval() instead.
 	*/
 	bool m_pending_removal = false;
```
Comment (SmallJoker) [inline src/server/serveractiveobject.h:231]: Could you please make this either protected or private, if possible?

File: src/server/serveractiveobject.h [hunk 3]
```diff
@@ -229,6 +236,7 @@ class ServerActiveObject : public ActiveObject
 
 		If this is set alongside with m_pending_removal, removal takes
 		priority.
+		Note: Do not assign this directly, use markForDeactivation() instead.
 	*/
 	bool m_pending_deactivation = false;
```
Comment (SmallJoker) [inline src/server/serveractiveobject.h:241]: Same here

File: src/server/serveractiveobject.h [hunk 4]
```diff
@@ -70,6 +70,12 @@ class ServerActiveObject : public ActiveObject
 	virtual bool environmentDeletes() const
 	{ return true; }
 
+	// Safely mark the object for removal or deactivation
+	void markForRemoval();
+	void markForDeactivation();
+	virtual void onMarkedForRemoval() {}
+	virtual void onMarkedForDeactivation() {}
```
Comment (SmallJoker) [inline src/server/serveractiveobject.h:77]: Should be protected

**Comment by SmallJoker** [comment]
```
2020-12-23 15:44:24: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
2020-12-23 15:44:24: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:25: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
2020-12-23 15:44:25: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:25: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
2020-12-23 15:44:25: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:26: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
2020-12-23 15:44:26: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
```

After `server_unload_unused_data_timeout = 10` seconds there's still no `on_deactivate`. Even waiting for 30s+ on a far distance made no difference.

File: src/server/serveractiveobject.h (hunk 2 above)
Comment (hecktest) [reply src/server/serveractiveobject.h:231]: This will require a `getThing() {return thing;}` which stinks but okay, I'll do it. C++ is a bad OOP language either way, there's no winning move here.

File: src/server/serveractiveobject.h (hunk 3 above)
Comment (hecktest) [reply src/server/serveractiveobject.h]: At least this won't require more dumb getters, nothing is asking for it directly.

**Comment by runsy** [comment]
> 
> 
> ```
> 2020-12-23 15:44:24: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
> 2020-12-23 15:44:24: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:25: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
> 2020-12-23 15:44:25: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:25: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
> 2020-12-23 15:44:25: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:26: ACTION[Server]: Callback entity: on_activate! pos=(-183,10,-88); dtime_s=0
> 2020-12-23 15:44:26: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> 2020-12-23 15:44:30: ACTION[Server]: Callback entity: get_staticdata! pos=(-183,10,-88)
> ```
> 
> After `server_unload_unused_data_timeout = 10` seconds there's still no `on_deactivate`. Even waiting for 30s+ on a far distance made no difference.
> 
> EDIT: Killing the entity works, though.

Normally mobs should deactivate when they are far away a distance of 80 more or less (the blocks are unloaded).

I have to test it anyway.

**Comment by hecktest** [comment]
Hang on, I'm working on it. The problem is that the definition of "deactivate" in code is somewhat confusing and that code path is never reached when an object is being frozen as static instead. (while thawing the object *does* call on_activate)

File: src/serverenvironment.cpp [hunk 5]
```diff
@@ -2068,6 +2069,7 @@ void ServerEnvironment::deactivateFarObjects(bool _force_delete)
 				force_delete = true;
 		}
 
+		obj->markForDeactivation();
```
Comment (hecktest) [inline src/serverenvironment.cpp:2072]: Let's try an experiment: Deactivate things in the function that was made for deactivating things. (wow!)
Please test.

I don't know what side effects this might bring about because the object lifecycle is just a big, big mess, but it seems to work.

File: src/server/serveractiveobject.cpp [hunk 6]
```diff
@@ -73,3 +73,21 @@ void ServerActiveObject::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &
...
+			onMarkedForRemoval();
+		m_pending_removal = true;
+	}
+}
+
+void ServerActiveObject::markForDeactivation()
+{
+	if (!m_pending_deactivation) {
```
Comment (SmallJoker) [inline src/server/serveractiveobject.cpp:88]: Is there a articular reason for this check? (same for code above)
`if (!isGone())` already does this, and a value assignment isn't expensive.

File: src/serverenvironment.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/serverenvironment.cpp:2072]: That does not sound very convincing, but the location seems right. However, this will need a comment like
`// Trigger the script callback after saving and prior deletion`

Will test.

**Comment by runsy** [comment]
I want to test it, but:

> Some checks were not successful

**Comment by SmallJoker** [comment]
Works. Will approve as soon the comments are handled.

File: src/server/serveractiveobject.cpp (hunk 6 above)
Comment (hecktest) [reply src/server/serveractiveobject.cpp:88]: Okay, so:
- There are two _competing_ deactivation paths, and possible interactions between them are pretty murky, this whole system isn't designed right. Lifecycles should probably be nested, not side by side.
- I needed to make sure the deactivation api is never ever called twice. This call checks whether the other path has been already invoked, without attempting to figure out if it ever could actually happen.
- Using `isGone` instead of the opposite path looks a bit cleaner, and since `isGone` is inline, the compiler ought to chip away the redundant part.

Basically, it was either this or reworking activation entirely and possibly making this too complex for a timely merge.

File: src/serverenvironment.cpp (hunk 5 above)
Comment (hecktest) [reply src/serverenvironment.cpp:2072]: Well I tried this and it worked, and it makes sense to deactivate things both when you delete and when you unload them.
The alternative was a monstrous condition along the lines of: _check if the object has been not really deactivated but in need of a deactivate call, and call the deactivate event through a special code path even thought we're not really deactivating._ That doesn't sound quite right. In other words, I'd have to add a third deactivation path, even though two are already too many.

I'm not sure what kind of comment could express the intent here properly, because this line of code has quite a history now:
- It was a bool assignment that among other things made isGone evaluate to true
- It got converted to a call that also tripped the deactivation, because deactivation should always happen before isGone is true
- There was still a code path where neither deactivation path would be called
- The line was moved out of its block to execute unconditionally because that actually makes a lot of sense in its new context

I'll think of something.

**Comment by hecktest** [comment]
Clang-format seems to hate my use of private: which I've done for a good reason. Like where does it... expect me to put that one function if an _inline_ public function then needs to call it? What a stupid rule.

File: src/server/serveractiveobject.cpp (hunk 6 above)
Comment (hecktest) [reply src/server/serveractiveobject.cpp:87]: Thought of a better way to do this, the redundant check now happens elsewhere but makes a **lot** more sense and also acts as a sanity check and documents something. (on_deactivate is guaranteed to always have a valid object state)

File: src/serverenvironment.cpp (hunk 5 above)
Comment (hecktest) [reply src/serverenvironment.cpp:2074]: Done, let me know if this is enough.

**Review by SmallJoker — APPROVED** [review]

**Review by sfan5 — APPROVED** [review]
Looks correct.

**Comment by runsy** [comment]
Thanz I will use it!
